### ISSUE TITLE
Some questions about “The committee machine: Computational to statistical gaps in learning a two-layers neural network”

### ISSUE THREAD

**Issue by Jimmy19970110** [benjaminaubin/TheCommitteeMachine#2]
Dear @benjaminaubin :
        I am a PhD student at the University of Chinese Academy of Sciences. I'm doing research on Bayesian deep learning.  I've been reading this paper of yours. The derivation of AMP algorithm in the paper is very detailed, which is of great help to me. Thanks a lot. There is one point in this paper that I never understood, and I don't seem to see the instructions in it. In section4.2 (Page 11), q = qd * IK×K + (qa/K)*1K*1K' , where does this formula come from? And what do q, qd and qa mean (the overlaps? I gauss )? If possible, could you give me some detailed description. Best wishes!

**Comment by benjaminaubin** [comment]
Dear Jimmy, 

Under the `committee symmetric` assumption in the Bayes-optimal setting, for W in R^{n x K}, the overlap matrix Q = lim_{n \to \infty} E[1/n * W' . W] is naturally parametrized by = (q_d) I_K×K + q_o 1_K . 1_K'   where q_d represents the diagonal term (in the case of non-specialized hidden units) of the overlap matrix, and q_o the off-diagonal term. 

I order to keep the spectrum of the overlap matrix Q of order 1, it requires to set q_o = O(1/K), so that we introduce q_a: q_o = q_a / K that yields Q = (q_d) I_K×K + q_a / K 1_K . 1_K'

Best!

**Comment by Jimmy19970110** [comment]
Dear @benjaminaubin :
Thank you for your reply. I've read your source code. When taking the generalization error, you use q to assign a value to qa and qd  (the source code : qa = self.K * self.q[0,1] ;   qd = self.q[0,0] - qa/self.K ). so the state evolution of training data and test data is the same? Besides, in supplementary material C : Generalization error, how does equation (79) get from equation (78)?  This bother me for a long time. Could you mind give me some detailed description about this? Best wishes!

**Comment by benjaminaubin** [comment]
Yes exactly Q = [[q00 , q01], [q01, q00]] = [[qd + qa/K, qa/K], [qa/K, qd + qa/K]] so that: qa/K = q01 and qd = q00 - qa/K = q00 -q01.

 Yet I do not understand your point "state evolution of training data and test data is the same": the state evolution of the AMP algorithm is controlled by the overlap parameter q. Its value depends on the number of examples seen during the training phase (called M=alpha * N): q(alpha). Then we compute the generalization error (the test error on the full distribution since we know it in our case, for a model trained with M = alpha * N examples) that depends only on this q(alpha) eq (11). 
 
 Concerning the explicit formulation of the generalization error eq (79) is simply the computation of the term E_X [\int_{R^4} Dz \sigma(\sigma(z1*) + \sigma(z2*)) * \sigma(\sigma(z1) + \sigma(z2)) ] that can be simplified to the extreme with a Cholesky decomposition of the overlap matrix with the form Q = (q_d) I_K×K + q_a / K 1_K . 1_K'.
 To help you with the computation please check the details of the computation in the attached notes. It leads to two expressions that are equivalent and must be evaluated with MC simulations in any case. 
 
[Generalization K=2.pdf](https://github.com/benjaminaubin/TheCommitteeMachine/files/5878886/Generalization.K.2.pdf)
[Generalization K=2_2.pdf](https://github.com/benjaminaubin/TheCommitteeMachine/files/5878887/Generalization.K.2_2.pdf)

I hope that all of this will help. Best!

**Comment by Jimmy19970110** [comment]
Dear @benjaminaubin 
Thank you for your sharing. I have read some other papers about AMP,and I found the definition of fw in your paper is different from others(I think its  miss this term---exp(-1/2*T^2*sigma^(-1))). So this term doesn't affect the accuracy of the AMP algorithm? Besides,in the derivation above Equation (107), 1/Z(mu-->i) = sqrt( det(A)/(2*pi)^K ), Is this an accurate derivation?
It's been bothering me for a long time. If possible, could you give me some detailed description.

This is another paper on AMP algorithm that I refer to.The definition of fw is eq.(31).
[https://arxiv.org/pdf/1910.00285.pdf](url)

 Best wishes!

**Comment by benjaminaubin** [comment]
About fw, the constant term can go outside the average over w, so that it disappears when you divide it by Zw: 
fw(T, Sigma) \equiv 1/Z_w \int d_w w P_w(w) exp(-1/2 * ( w - T)^2 Sigma^{-1}) = \frac{\int d_w w P_w(w) exp(-1/2 * ( w - T)^2 Sigma^{-1}}{\int d_w P_w(w) exp(-1/2 * ( w - T)^2 Sigma^{-1})} = \frac{\int d_w w P_w(w) exp(-1/2 * w^2 Sigma^{-1} + w Sigma^{-1} T )}{\int d_w P_w(w) exp(-1/2 * w^2 Sigma^{-1} + w Sigma^{-1} T)}
so that both definitions coincide.

About eq 107, there should be a \simeq above eq 107 in the last line, since we sum the expansion of the exponential. And since we would like the distribution to be normalized, the normalization (that we do not care about) has to be sqrt( det(A)/(2*pi)^K ) after resuming the expansion to a Gaussian. 

Best
